File: 98.validate-binary-search-tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        MIN_VALUE = -2**31 - 1
        MAX_VALUE = 2**31
        return self.validate(root, MIN_VALUE, MAX_VALUE)

    def validate(self, node: TreeNode, minVal: int, maxVal: int) -> bool:
        leftValidation = rightValidation = True

        if node:
            if node.val <= minVal or node.val >= maxVal:
                return False

            leftValidation = self.validate(node.left, minVal, node.val)
            rightValidation = self.validate(node.right, node.val, maxVal)

        return leftValidation and rightValidation

# @lc code=end

File: 98.validate-binary-search-tree.js
/*
 * @lc app=leetcode id=98 lang=javascript
 *
 * [98] Validate Binary Search Tree
 */

// Definition for a binary tree node.
class TreeNode { 
  constructor(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
  }
}

// @lc code=start
/**
 * @param {TreeNode} node
 * @param {number} left
 * @param {number} right
 * @return {boolean}
 */
const validate = (node, left, right) => {
  if (node === null) {
    return true;
  }

  if (node.val > left && node.val < right) {
    return (validate(node.left, left, node.val)
      && validate(node.right, node.val, right))
  } else {
    return false;
  }
};

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isValidBST = (root) => {
  return validate(root, Math.pow(-2, 31) - 1, Math.pow(2, 31));
};

// @lc code=end

